### DISCUSSION TITLE
Time weighted average?

### DISCUSSION THREAD

**Discussion by abh** [ClickHouse/ClickHouse#55582] [General]
I am logging time series type data at irregular intervals. Is there a reasonable way to take data like

| series | timestamp                       | value |
|------|-----------------------| ------|
|  1         | 2023-10-14 00:10:11  | 5 |
|  1         | 2023-10-14 00:14:20 | 10 |
|  1         | 2023-10-14 00:31:05 | 8 |
| ... |

and calculate an averaged weighted by the time at each state?

**Comment by den-crane** [comment]
What is averaged weighted? What result do you expect?

https://fiddle.clickhouse.com/2b16096f-4f04-49e3-851f-71693528313e

```sql
CREATE TABLE series (series Int64,timestamp DateTime , value Float64) ENGINE=Memory
as select * from values(
(1, '2023-10-14 00:10:11',	5),
(1, '2023-10-14 00:14:20', 	10),
(1, '2023-10-14 00:31:05', 	8));

select series, timestamp, value, time_diff,
      round(value / time_diff,4) res
from (
select series, 
  arraySort(i->i.1, groupArray((timestamp, value))) ga,
  arrayDifference(ga.1) diff
from series
group by series)
array join ga.1 as timestamp, ga.2 as value, diff as time_diff
order by series, timestamp
format PrettyCompactMonoBlock

+-series-+-----------timestamp-+-value-+-time_diff-+----res-+
|      1 | 2023-10-14 00:10:11 |     5 |         0 |    inf |
|      1 | 2023-10-14 00:14:20 |    10 |       249 | 0.0402 |
|      1 | 2023-10-14 00:31:05 |     8 |      1005 |  0.008 |
+--------+---------------------+-------+-----------+--------+
```

**Comment by KnutZuidema** [comment]
@abh 

Old discussion here, but to calculate a time-weighted average one needs to choose some kind of interpolation method. Using the interpolated values and a time-weight (i.e. the diff between two adjacent observation timestamps) one can use the `avgWeighted` function provided by clickhouse to calculate a time-weighted average.

That might look like this when using linear interpolation:

```sql
WITH series AS (
    SELECT * FROM values('id Int, timestamp DateTime64, value Nullable(Float64)',
(1, '2023-10-14 00:10:11', 5),
(1, '2023-10-14 00:14:20', 10),
(1, '2023-10-14 00:31:05', 8))
),
interpolated_series AS ( 
  SELECT
    id,
    dateDiff('second', lag(timestamp) OVER window, timestamp) time_weight,
    (lag(value) OVER window + value) / 2 AS interp_value
  FROM series
  WINDOW window AS (PARTITION BY id ORDER BY timestamp)
)
SELECT id, avgWeighted(interp_value, time_weight) AS time_weighted_average
FROM interpolated_series
GROUP BY id;
```

I would love to see something like this implemented natively in clickhouse with potential for more sophisticated interpolation methods. TimescaleDB does offer such an aggregation function, see [here](https://www.tigerdata.com/docs/use-timescale/latest/hyperfunctions/time-weighted-averages/time-weighted-average).
I'd volunteer to do this, however I'd need guidance from somebody more experienced with the ClickHouse codebase, as my C++ is quite rusty.
